### PR TITLE
fix(604): Constructor takes in a config object

### PR DESCRIPTION
**Pull request by tkyi**
## Context
We are changing how the executor section will be configured in the Screwdriver API.yaml.

The API.yaml will look like:
```yaml
executor:
   defaultPlugin: docker
   plugins:
    - name: k8s
      options:
        kubernetes:
            host: kubernetes.default
        launchVersion: stable
    - name: docker
      options:
        docker: {}
        launchVersion: stable
```
To construct an ExecutorRouter object:
```
const executorConfig = config.get('executor');
const ExecutorRouter = require('screwdriver-executor-router');
const executor = new ExecutorRouter({ defaultPlugin: executorConfig.defaultPlugin, executor: executorConfig.plugins, ecosystem });
```

## Related Links
Follow up to https://github.com/screwdriver-cd/executor-router/pull/5
Related to https://github.com/screwdriver-cd/screwdriver/issues/604

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,34 +1,39 @@
...
+        const ecosystem = config.ecosystem || {};
+        const executorConfig = config.executor;
+        const registrations = (Array.isArray(executorConfig)) ? executorConfig : [executorConfig];
 
         registrations.forEach((plugin, index) => {
             // eslint-disable-next-line global-require, import/no-dynamic-require
             const ExecutorPlugin = require(plugin.name);
-
-            const executorPlugin = new ExecutorPlugin(plugin.options);
+            const options = hoek.applyToDefaults({ ecosystem }, plugin.options); // Add ecosystem to executor options
```
Comment (ian-fox) [inline index.js]: `applyToDefaults` is can also be done without having to `include` hoek with [`Object.assign`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign)

File: index.js [hunk 2]
```diff
@@ -1,34 +1,38 @@
...
+            throw new Error('No executor config passed in.');
+        }
+
         super();
 
-        const plugins = options.plugins;
-        const registrations = (Array.isArray(plugins)) ? plugins : [plugins];
+        const ecosystem = config.ecosystem || {};
+        const executorConfig = config.executor;
+        const registrations = (Array.isArray(executorConfig)) ? executorConfig : [executorConfig];
```
Comment (petey) [inline index.js]: Should probably throw if the array is empty.

File: index.js [hunk 3]
```diff
@@ -1,34 +1,38 @@
...
 
         registrations.forEach((plugin, index) => {
             // eslint-disable-next-line global-require, import/no-dynamic-require
             const ExecutorPlugin = require(plugin.name);
```
Comment (petey) [inline index.js]: Throws when requiring a module that doesn't exist.

File: index.js (hunk 2 above)
Comment (tkyi) [inline index.js]: How do I figure out if there's an invalid element?

File: index.js (hunk 3 above)
Comment (tkyi) [inline index.js]: Wait do you want me to throw it or not throw it?

File: index.js [hunk 4]
```diff
@@ -1,34 +1,48 @@
...
-        const registrations = (Array.isArray(plugins)) ? plugins : [plugins];
+        executorConfig.forEach((plugin, index) => {
+            let ExecutorPlugin;
 
-        registrations.forEach((plugin, index) => {
-            // eslint-disable-next-line global-require, import/no-dynamic-require
-            const ExecutorPlugin = require(plugin.name);
+            try {
+                // eslint-disable-next-line global-require, import/no-dynamic-require
+                ExecutorPlugin = require(`screwdriver-executor-${plugin.name}`);
```
Comment (petey) [inline index.js]: hmm. I thought they were providing the full plugin name already.

File: index.js [hunk 5]
```diff
@@ -1,34 +1,48 @@
...
 
-            const executorPlugin = new ExecutorPlugin(plugin.options);
+            const options = Object.assign({ ecosystem }, plugin.options); // Add ecosystem to executor options
+            const executorPlugin = new ExecutorPlugin(options);
 
             // Set the default executor
-            if (index === 0) {
-                this.default_executor = executorPlugin;
+            if (index === 0 && !config.defaultPlugin) {
+                this.defaultExecutor = executorPlugin;
```
Comment (petey) [inline index.js]: This module should probably throw if it does not have a valid defaultExecutor, after everything is initialized.

File: index.js (hunk 4 above)
Comment (tkyi) [inline index.js]: No we changed it back for backwards compatibility with sd-in-a-box.py

**Pull request merged**